8351319: AOT cache support for custom class loaders broken since JDK-8348426

iklam opened the pull request:
Since [JDK-8348426](https://bugs.openjdk.org/browse/JDK-8348426) (Generate binary file for -XX:AOTMode=record -XX:AOTConfiguration=file), the AOT cache no longer contains classes intended for custom class loaders (these are called "unregistered classes" in CDS terminology).

The fix is simple -- we already remember the set of unregistered classes in the AOT configuration file. We just need to add them into the final AOT cache (see changes in finalImageRecipes.cpp).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8351319](https://bugs.openjdk.org/browse/JDK-8351319): AOT cache support for custom class loaders broken since JDK-8348426 (**Bug** - P2)


### Reviewers
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**)
 * [Matias Saavedra Silva](https://openjdk.org/census#matsaave) (@matias9927 - **Reviewer**) Review applies to [ccea0a41](https://git.openjdk.org/jdk/pull/23926/files/ccea0a412f44f735fb8f8606a3db054ea5bdda1e)
 * [John R Rose](https://openjdk.org/census#jrose) (@rose00 - **Reviewer**) Review applies to [ccea0a41](https://git.openjdk.org/jdk/pull/23926/files/ccea0a412f44f735fb8f8606a3db054ea5bdda1e)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/23926/head:pull/23926` \
`$ git checkout pull/23926`

Update a local copy of the PR: \
`$ git checkout pull/23926` \
`$ git pull https://git.openjdk.org/jdk.git pull/23926/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 23926`

View PR using the GUI difftool: \
`$ git pr show -t 23926`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/23926.diff">https://git.openjdk.org/jdk/pull/23926.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/23926#issuecomment-2702751768)
</details>

calvinccheung reviewed (approved):
LGTM

matias9927 reviewed (approved):
LGTM!

rose00 reviewed (approved):
Good.

I suppose there already tests for the other end of the process, where an unregistered class in the AOT cache is actually used.  What are those tests?

matias9927 reviewed (approved):
Changes look good!

iklam commented:
> Good.
> 
> I suppose there already tests for the other end of the process, where an unregistered class in the AOT cache is actually used. What are those tests?

It turns out that the existing tests only check for unregistered classes in the classical CDS archive. I added a test in BulkLoaderTest.java to check the AOT cache and found a bug. Fixed in [c02a611](https://github.com/openjdk/jdk/pull/23926/commits/c02a611ecca7da318865d3d98e4cae7fb1eb8410)

iklam commented:
Thanks @calvinccheung @matias9927 @rose00 for the review
/integrate
